File: src/main/java/com/atsuishio/superbwarfare/client/renderer/animations/SenpaiAnimationInstance.java
package com.atsuishio.superbwarfare.client.renderer.animations;

import com.atsuishio.superbwarfare.entity.SenpaiEntity;
import com.maydaymemory.mae.basic.Pose;
import com.maydaymemory.mae.control.statemachine.AnimationStateMachine;

public class SenpaiAnimationInstance {
    private final SenpaiContext context;
    private final AnimationStateMachine<SenpaiContext> stateMachine;

    public SenpaiAnimationInstance(SenpaiEntity entity) {
        this.context = new SenpaiContext(entity);
        this.stateMachine = new AnimationStateMachine<>(SenpaiStates.IDLE, context, System::nanoTime);
        SenpaiStates.IDLE.onEnter(this.context, SenpaiStates.IDLE);
    }

    public void tick() {
        stateMachine.tick();
        context.tick();
    }

    public SenpaiContext getContext() {
        return context;
    }

    public Pose getPose() {
        return stateMachine.getPose();
    }
}

File: src/main/kotlin/com/atsuishio/superbwarfare/client/animation/entity/SenpaiAnimationInstance.kt
package com.atsuishio.superbwarfare.client.animation.entity

import com.atsuishio.superbwarfare.entity.SenpaiEntity
import com.maydaymemory.mae.basic.Pose
import com.maydaymemory.mae.control.statemachine.AnimationStateMachine

class SenpaiAnimationInstance(entity: SenpaiEntity) {
    val context: SenpaiContext = SenpaiContext(entity)
    private val stateMachine = AnimationStateMachine(SenpaiStates.IDLE, context) { System.nanoTime() }

    init {
        SenpaiStates.IDLE.onEnter(this.context, SenpaiStates.IDLE)
    }

    fun tick() {
        stateMachine.tick()
        context.tick()
    }

    fun getPose(): Pose {
        return stateMachine.getPose()
    }
}
